esutil/bulk_indexer.go

```diff
@@ -0,0 +1,509 @@
+// Licensed to Elasticsearch B.V. under one or more agreements.
+// Elasticsearch B.V. licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+package esutil
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/elastic/go-elasticsearch/v8"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+// BulkIndexer represents a parallel, asynchronous, efficient indexer for Elasticsearch.
+//
+type BulkIndexer interface {
+	// Add adds an item to the indexer. It returns an error when the item cannot be added.
+	// Use the OnSuccess and OnFailure callbacks to get the operation result for the item.
+	//
+	// You must call the Close() method after you're done adding items.
+	//
+	// It is safe for concurrent use. When it's called from goroutines,
+	// they must finish before the call to Close, eg. using sync.WaitGroup.
+	Add(context.Context, BulkIndexerItem) error
+
+	// Close waits until all added items are flushed and closes the indexer.
+	Close(context.Context) error
```

Comment:
I wonder if the behavior of close overlaps with context cancellation and wait groups. Do I need to cancel added items if I want close to return early? If cancelled, will my callback be run?

Maybe it would make sense to have 'Close' on the concrete type only, but remove it from the interface?

Reply:
I went through several different names, including `Wait()`, but finally decided on `Close()`. It does not "close" the indexer in a sense of "drop what you're doing and shut down", but rather signals to the indexer that it should finish the in-flight operations and close itself for more.

> Do I need to cancel added items if I want close to return early?

Well, currently there's no semantics of returning early. As `Close()` takes a context, I imagine that  propagating eg. a timeout to the `worker.flush()` method would be possible — but I have focused only on making it possible, not implementing it.

> If cancelled, will my callback be run?

Currently, the callback should be called for every added item — also in case of cancel/timeout... But I would have to write some code to test it.

Reply:
I see. I think some early cancel + shutdown is a must. If the library is used by a daemon, then daemon restart or shutdown should not be blocked by an unresponsive network or Elasticsearch instance.

Having a context.Context on Close for cancellation just feels a little weird.

Reply:
Right! I'll have a look into propagating the errors, now that it's been made possible by 9ec0d6b. As @axw suggested above, `flush()` must take a context, and then it should be possible to pass down the context with timeout from `Close()` into it.

Reply:
@urso, I've added a `done` channel to the indexer, which is notified in `Close()` in d294483. I've verified with https://github.com/uber-go/goleak in the corresponding test added in 9cb7b6a that it doesn't leak a goroutine. I assume, however, that since `worker.run()` is not currently able to return early, that it will block closing or leak a goroutine. Maybe that's something we can chat about.

Reply:
Given that we add more callbacks, how about an 'OnCancelled' callback, that allows me to capture events that have not been published yet when I shutdown the ES output. Use-case is again reconfiguration:
1. block publishing new events 
2. create new BulkIndexer
3. close old BulkdIndexer
4. forward all cancelled events to the new BulkIndexer
5. unblock publishing new events.